static/js/mld_store.js

```diff
@@ -0,0 +1,84 @@
+"use strict";
+
+class MldStore {
+    constructor() {
+        this._objects = new Map();
+    }
+
+    add(mld, key) {
+        if (key === undefined) {
+            key = hash_util.operators_to_hash(mld.filter.operators());
+        }
+        this._objects.set(key, mld);
+    }
+
+    get_key(mld) {
+        if (!mld || !mld.filter || mld.filter.operators().length === 0) {
+            return undefined;
+        }
+
+        return hash_util.operators_to_hash(mld.filter.operators());
+    }
+
+    get(key) {
+        return this._objects.get(key);
+    }
+
+    get_key_from_operators(raw_operators) {
+        const filter = new Filter(raw_operators);
+        const operators = filter.operators();
+        return hash_util.operators_to_hash(operators);
+    }
+
+    get_from_operators(raw_operators) {
+        const key = this.get_key_from_operators(raw_operators);
+        return this.get(key);
+    }
+
+    get_valid_mlds(filter) {
+        const operators = filter.operators();
+        const key = hash_util.operators_to_hash(operators);
+        const keys = [key];
+
+        if (filter.contains_only_private_messages()) {
+            if (keys[0].includes("pm-with") || keys[0].includes("group-pm-with")) {
+                const is_private_operator = [{operator: "is", operand: "private"}];
+                keys.push(this.get_key_from_operators(is_private_operator));
+            }
+        } else {
+            if (keys[0].includes("topic")) {
+                const stream_operator = [filter.operators()[0]];
+                keys.push(this.get_key_from_operators(stream_operator));
+            }
+        }
+
+        const mld_objects = keys.map((key) => this.get(key));
+        return mld_objects.filter((mld) => mld !== undefined);
+    }
+
+    has(key) {
+        return this._objects.has(key);
+    }
+
+    delete(key) {
+        this._objects.delete(key);
+    }
+
+    keys() {
+        return Array.from(this._objects.keys());
+    }
+
+    empty() {
+        this._objects = new Map();
+    }
+
+    forEach(f) {
+        this._objects.forEach(f);
+    }
+}
+
+const stored_mld = new MldStore();
+exports.MldStore = MldStore;
+
+window.stored_mld = stored_mld;
+window.mld_store = exports;
```

Comment:
Let's do this as a proper module, rather than using the `window` hack that we're trying to purge from the codebase.  The code under `static/shared` has some good examples of modern modules.  @andersk FYI to help with this.

I'd also like to think about naming for this module; I think `message_list_data_cache.js` might be a good name?  The `cache` part specifies that the intent/structure is that we can discard any of these that isn't in use for rendering at any time (either to save memory or because we aren't sure to how to update it properly) without introducing correctness issues.